docs: add directives page and update docs nav (#10885)
apollographql/apollo-client: docs/source/data/directives.mdx

  person(id: $personId) {
    # Basic fields (fast)
    id
    firstName
    lastName

    # Friend fields (slower)
    ... @defer {
      friends {
        id
      }
    }
  }
}
```

For more information about the `@defer` directive, check out the [`@defer` docs](/react/data/defer).

## `@export`

If your GraphQL query uses variables, the local-only fields of that query can provide the values of those variables.

To do so, you apply the `@export(as: "variableName")` directive, like so:

```js
const GET_CURRENT_AUTHOR_POST_COUNT = gql`
  query CurrentAuthorPostCount($authorId: Int!) {
    currentAuthorId @client @export(as: "authorId")
    postCount(authorId: $authorId)
  }
`;
```

In the query above, the result of the local-only field `currentAuthorId` is used as the value of the `$authorId` variable that's passed to `postCount`.

You can do this even if `postCount` is also a local-only field (i.e., if it's also marked as `@client`).

For more information and other considerations when using the `@export` directive, check out the [local-only fields docs](/react/local-state/managing-state-with-field-policies/#using-local-only-fields-as-graphql-variables).

> ## `@nonreactive`

Reviewer: This section is 👨‍🍳 💋

    name
    status
  }
`;

const ALL_TRAILS = gql`
  query allTrails {
    allTrails {
      id
      ...TrailFragment @nonreactive
    }
  }
  ${TrailFragment}
`;

function App() {
  const { data, loading } = useQuery(ALL_TRAILS);
  return (
    <main>
      <h2>Ski Trails</h2>
      <ul>
        {data?.trails.map((trail) => (
          <Trail key={trail.id} id={trail.id} />
        ))}
      </ul>
    </main>
  );
}
```

The `Trail` component renders a trail's name and status and allows the user to execute a mutation to toggle the status of the trail between `"OPEN"` and `"CLOSED"`:

```jsx
const Trail = ({ id }) => {
  const [updateTrail] = useMutation(UPDATE_TRAIL);
  const { data } = useFragment({
    fragment: TrailFragment,
    from: {
      __typename: "Trail",
      id,
    },
  });
  return (
    <li key={id}>
      {data.name} - {data.status}
      <input
        checked={data.status === "OPEN" ? true : false}
        type="checkbox"
        onChange={(e) => {
          updateTrail({
            variables: {
              trailId: id,
              status: e.target.checked ? "OPEN" : "CLOSED",
            },
          });
        }}
      />
    </li>
  );
};
```

> Notice that the `Trail` component isn't receiving the entire `trail` object via props, only the `id` which is used along with the fragment document to create a live binding for each trail item in the cache. This allows each `Trail` component to react to the cache updates for a single trail independently, without the entire list re-rendering. Updates to a trail's `status` will not cause the parent `App` component to rerender given the `@nonreactive` directive applied to the `TrailFragment` spread, a fragment which includes the `status` field.

Reviewer: ```suggestion
Notice that the `Trail` component isn't receiving the entire `trail` object via props, only the `id` which is used along with the fragment document to create a live binding for each trail item in the cache. This allows each `Trail` component to react to the cache updates for a single trail independently, without the entire list rerendering. Updates to a trail's `status` will not cause the parent `App` component to rerender given the `@nonreactive` directive applied to the `TrailFragment` spread, a fragment which includes the `status` field.
```

Removed the dash since other uses of this word don't include it 🙂

> Notice that the `Trail` component isn't receiving the entire `trail` object via props, only the `id` which is used along with the fragment document to create a live binding for each trail item in the cache. This allows each `Trail` component to react to the cache updates for a single trail independently, without the entire list re-rendering. Updates to a trail's `status` will not cause the parent `App` component to rerender given the `@nonreactive` directive applied to the `TrailFragment` spread, a fragment which includes the `status` field.

Reviewer: The last sentence in this paragraph feels a bit long and slightly repetetive with the previous sentence. What about something like this?

```suggestion
Notice that the `Trail` component isn't receiving the entire `trail` object via props, only the `id` which is used along with the fragment document to create a live binding for each trail item in the cache. This allows each `Trail` component to react to the cache updates for a single trail independently. Updates to a trail's `status` will not cause the parent `App` component to rerender since it applies the `@nonreactive` directive to the `TrailFragment` spread; a fragment which includes the `status` field.
```

---
title: "Using GraphQL directives in Apollo Client"
description: Configure GraphQL fields and fragments
---

A directive decorates part of a GraphQL schema or operation with additional configuration. Tools like Apollo Client can read a GraphQL document's directives and perform custom logic as appropriate.

Directives are preceded by the `@` character, like so:

```graphql
query myQuery($someTest: Boolean) {
  experimentalField @skip(if: $someTest)
}
```

> This example shows the `@skip` directive, which is a default directive (i.e., it's part of the [GraphQL specification](http://spec.graphql.org/June2018/#sec--deprecated)). It demonstrates the following about directives:

Author: Thanks for the review! The `default directive` term is from the AS directives page: https://www.apollographql.com/docs/apollo-server/schema/directives/ - I was also hoping to find the term in the spec itself and defaulted to staying consistent, but I also prefer built-in. Missed updating that link, thanks for that catch!

> This example shows the `@skip` directive, which is a default directive (i.e., it's part of the [GraphQL specification](http://spec.graphql.org/June2018/#sec--deprecated)). It demonstrates the following about directives:

Reviewer: Ahhhhh ok. I'm fine with the consistency in our own docs so totally fine to leave `default directive` if you want to. I'm happy either way. Thanks for pointing this out!

- Directives can take arguments of their own (`if` in this case).
- Directives appear after the declaration of what they decorate (the `experimentalField` field in this case).

## `@client`

The `@client` directive tells Apollo Client _not_ to send the field(s) it is annotating to the server.

```graphql
query LaunchDetails($launchId: ID!) {
  launch(id: $launchId) {
    site
    rocket {
      type
      # populated via the read function configured on InMemoryCache's typePolicies,
      # never sent with the request to the server
      description @client
    }
  }
}
```

> For more information about the `@client` directive, see this section on [local-only fields](/react/local-state/managing-state-with-field-policies/). The `@client` directive is also useful for [client schema mocking](/react/development-testing/client-schema-mocking/#3-query-with-the-client-directive) before a given field is supported in the GraphQL API your application is consuming.

Author: The main issue I see with linking to the overview page is that it only has one small mention of `@client` on it, vs. the field policies doc which is our best step-by-step guide to using `@client`. I can add a mention of local resolvers, but I also think it's fine to omit given that this is the directives page, and our local resolvers doc still instructs users to prefer field policies at the top: https://www.apollographql.com/docs/react/local-state/local-resolvers

Thoughts?

> For more information about the `@client` directive, see this section on [local-only fields](/react/local-state/managing-state-with-field-policies/). The `@client` directive is also useful for [client schema mocking](/react/development-testing/client-schema-mocking/#3-query-with-the-client-directive) before a given field is supported in the GraphQL API your application is consuming.

Reviewer: I can live with that. Thanks for talking this through!

## `@connection`

> - /react/caching/advanced-topics/#the-connection-directive

Author: Meant to add a note here :) My comment was going to be: do we want this in the directives docs given `keyArgs`? But based on your comment it seems like this would be good to include, so I'll move forward with your recommendation 👍

> - /react/caching/advanced-topics/#the-connection-directive

Author: https://github.com/apollographql/apollo-client/pull/10885/commits/8daea0430d79573ee36f7312f844aa3e0ca4a439

There's a lot going on in the example query, so for this one I'm of the mind it's better to leave out here and just let the linked page do the explaining.

> - /react/caching/advanced-topics/#the-connection-directive

Reviewer: Do you think we could simplify the example perhaps? I find examples, even something that might look complex, go a long way to understand at-a-glance what to expect. Its the "picture is worth 1000 words" scenario. This page doesn't necessarily need to detail the ins and outs of the directive, or all the arguments it takes, but a single code sample I find is better than nothing as you can usually grok the purpose much quicker.

What about something like this? 
```graphql
query Feed($offset: Int, $limit: Int) {
  feed(offset: $offset, limit: $limit) @connection(key: "feed") {
    ...FeedFields
  }
}
```

This feels like it communicates the pagination aspect of it while keeping it a bit simpler.

Thoughts?

> - /react/caching/advanced-topics/#the-connection-directive

Author: Sure, it can't hurt to add a simplified snippet: [3845df4](https://github.com/apollographql/apollo-client/pull/10885/commits/3845df4221d0ec2497f7969f1a7bee51d184dc2c)

> - /react/caching/advanced-topics/#the-connection-directive

Reviewer: Awesome. Thank you so so much ❤️

## `@defer`

> Beginning with version `3.7.0`, Apollo Client Web provides preview support for [the `@defer` directive](https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md). This directive enables your queries to receive data for specific fields _incrementally_, instead of receiving all field data at the same time. This is helpful whenever some fields in a query take much longer to resolve than others.

Reviewer: Totally understand, but I'd argue the existing docs could also use this update. For example, our first 3 pages in the docs market the name as "Apollo Client", not "Apollo Client Web"

- https://www.apollographql.com/docs/react/
- https://www.apollographql.com/docs/react/why-apollo
- https://www.apollographql.com/docs/react/get-started

> Beginning with version `3.7.0`, Apollo Client Web provides preview support for [the `@defer` directive](https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md). This directive enables your queries to receive data for specific fields _incrementally_, instead of receiving all field data at the same time. This is helpful whenever some fields in a query take much longer to resolve than others.

Author: Sure, no problem, updated all references in our docs: [0d6046b](https://github.com/apollographql/apollo-client/pull/10885/commits/0d6046bed759ce065711e52392ead322d2685ea8)

There are still references to Apollo Client Web in other docs https://www.apollographql.com/docs/apollo-server/data/subscriptions/

> Beginning with version `3.7.0`, Apollo Client Web provides preview support for [the `@defer` directive](https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md). This directive enables your queries to receive data for specific fields _incrementally_, instead of receiving all field data at the same time. This is helpful whenever some fields in a query take much longer to resolve than others.

Reviewer: I think cross linking from another part of the Apollo docs outside of the client to a client web doc is ok since you're outside the context of the client docs (at least I think so). Thanks so much for updating this!

To use the `@defer` directive, we apply it to a field or fragment (in-line or named) that contains all slow-resolving fields:

```graphql
query PersonQuery($personId: ID!) {
  person(id: $personId) {

>     # Basic fields (fast)
>     id
>     firstName
>     lastName
> 
>     # Friend fields (slower)

Reviewer: ```suggestion
    # Basic fields
    id
    firstName
    lastName

    # Friend fields (deferred)
```

Same here. You might consider the term "deferred" here

Author: Same with this example (shared with defer doc)

    ... @defer {
      friends {
        id
      }
    }
  }
}
```

For more information about the `@defer` directive, check out the [`@defer` docs](/react/data/defer).

## `@export`

If your GraphQL query uses variables, the local-only fields of that query can provide the values of those variables.

To do so, you apply the `@export(as: "variableName")` directive, like so:

```js
const GET_CURRENT_AUTHOR_POST_COUNT = gql`
  query CurrentAuthorPostCount($authorId: Int!) {
    currentAuthorId @client @export(as: "authorId")
    postCount(authorId: $authorId)
  }
`;
```

In the query above, the result of the local-only field `currentAuthorId` is used as the value of the `$authorId` variable that's passed to `postCount`.

You can do this even if `postCount` is also a local-only field (i.e., if it's also marked as `@client`).

> For more information about the `@export` directive and other considerations when using `@export`, check out the [local-only fields docs](/react/local-state/managing-state-with-field-policies/#using-local-only-fields-as-graphql-variables).

Reviewer: ```suggestion
For more information about the `@export` directive and its other considerations, check out the [local-only fields docs](/react/local-state/managing-state-with-field-policies/#using-local-only-fields-as-graphql-variables).
```

I think you can simplify this sentence by using "its" to refer back to the `@export` directive rather than restating it.

Author: https://github.com/apollographql/apollo-client/pull/10885/commits/8daea0430d79573ee36f7312f844aa3e0ca4a439
